elastic/kibana
# chore(slo): Compute SLI value and error budget consumption

## 📝  Summary

Resolves https://github.com/elastic/kibana/issues/142521

💬 Suggest to review commit per commit 

This PR introduces a SLIClient used to fetch the current SLI data of the SLO. It currently only supports SLO defined with an occurrences' budgeting method and a rolling time window (since we don't support other type of SLO at the moment).

I've introduced two domain services to encapsulate the domain logic, using only other domain models. Usually this logic would have been defined directly within the domain models, but we don't have any (we only have types), thus the functional approach used here. 

I've made some arbitrary decision here:
- Error consumed can go higher than 1 (i.e. 100%), e.g. You could have consumed 3,000% of your error budget.
- Error left cannot go below 0 (i.e. 0%), e.g. When you have consumed 3,000% of your error budget, you have 0% left.
- When the `good >= total` events, we consider that `good = total`, therefore error budget consumption is 0% and left is 100%


## ✅  Manual testing

Reuse the query from https://github.com/elastic/kibana/pull/142273 to create and retrieve an SLO. The GET api should include a `summary` field with the current sli value as well as the error budget consumption.

1. Create an SLO with some wrong indicator parameters, e.g. you'll get no sli data
2. Create an SLO with good indicator parameters, e.g. you'll get sli data

<details>
<summary>👇 Bad indicator parameters (params.environment), yielding no SLI data</summary>

```
{
	"id": "67aa3d10-44db-11ed-b55e-a7742e4854bf",
	"name": "My SLO Availability",
	"description": "99% o11y-app all services availablility",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "inexistant",
			"service": "o11y-app",
			"transaction_type": "request",
			"transaction_name": "GET /flaky",
			"good_status_codes": [
				"2xx",
				"3xx",
				"4xx"
			]
		}
	},
	"time_window": {
		"duration": "7d",
		"is_rolling": true
	},
	"budgeting_method": "occurrences",
	"objective": {
		"target": 0.99
	},
	"summary": {
		"sli_value": 0,
		"error_budget": {
			"initial": 0.01,
			"consumed": 0,
			"remaining": 1
		}
	},
	"revision": 1,
	"created_at": "2022-10-05T18:27:44.609Z",
	"updated_at": "2022-10-05T18:27:44.609Z"
}
```
</details>

<details>
<summary>👇 Good indicator parameters, yielding SLI data</summary>

```
{
	"id": "fb7bea40-43e4-11ed-aa34-af16f78c8a81",
	"name": "My SLO Availability",
	"description": "99% o11y-app all services availablility",
	"indicator": {
		"type": "slo.apm.transaction_error_rate",
		"params": {
			"environment": "development",
			"service": "o11y-app",
			"transaction_type": "request",
			"transaction_name": "GET /flaky",
			"good_status_codes": [
				"2xx",
				"3xx",
				"4xx"
			]
		}
	},
	"time_window": {
		"duration": "7d",
		"is_rolling": true
	},
	"budgeting_method": "occurrences",
	"objective": {
		"target": 0.95
	},
	"summary": {
		"sli_value": 0.998756,
		"error_budget": {
			"initial": 0.05,
			"consumed": 0.024872,
			"remaining": 0.975128
		}
	},
	"revision": 1,
	"created_at": "2022-10-04T13:03:46.916Z",
	"updated_at": "2022-10-04T13:03:46.916Z"
}
```
</details>
